File: src/console/command/Command.php
<?php

namespace pocketcloud\cloud\console\command;

use pocketcloud\cloud\console\command\parameter\exception\ArgumentParseException;
use pocketcloud\cloud\console\command\parameter\exception\FlagParseException;
use pocketcloud\cloud\console\command\parameter\exception\NoArgumentFoundException;
use pocketcloud\cloud\console\command\sender\ICommandSender;
use pocketcloud\cloud\console\command\util\CommandFlagTrait;
use pocketcloud\cloud\console\command\util\CommandParameterTrait;
use pocketcloud\cloud\console\log\level\CloudLogLevel;
use pocketcloud\cloud\util\promise\Promise;

abstract class Command {
    use CommandParameterTrait, CommandFlagTrait;

    /** @var array<SubCommand> */
    private array $subCommands = [];

    private bool $useRegularHandlerForSubCommands = false;

    public function __construct(
        private readonly string $name,
        private readonly string $description,
        private readonly array $aliases = [],
        private readonly ?string $usage = null
    ) {}

    final public function waitForConfirmation(ICommandSender $sender, string $prompt, array $keywordsAccept, int $timeout = 10): Promise {
        return CommandManager::getInstance()->waitForConfirmation($this, $sender, $prompt, $keywordsAccept, $timeout);
    }

    public function enableUseRegularHandlerForSubCommands(bool $val = true): self {
        $this->useRegularHandlerForSubCommands = $val;
        return $this;
    }

    /** @internal */
    public function handle(ICommandSender $sender, string $label, array $args): void {
        try {
            ["globalFlags" => $commandGlobalFlags, "regularFlags" => $commandRegularFlags] = $this->scanAndCleanFlags($args);
        } catch (FlagParseException) {
            $this->sendUsageMessage($sender);
            return;
        }

        $subCommand = null;
        if (!empty($this->subCommands)) {
            if ($this->mustUseSubCommands() && count($args) < 1) {
                $this->sendUsageMessage($sender);
                return;
            }

            if (count($args) > 0) $subCommand = $this->getSubCommand($args[0]);
            if ($subCommand === null && $this->mustUseSubCommands()) {
                $this->sendUsageMessage($sender);
                return;
            } else if ($subCommand instanceof SubCommand) array_shift($args);
        }

        $subCommandFlags = [];
        $currentParameter = null;
        try {
            if ($subCommand === null) {
                $parsedArgs = $this->parseArgs($args, $currentParameter);
            } else {
                ["globalFlags" => $_, "regularFlags" => $subCommandFlags] = $subCommand->scanAndCleanFlags($args, true);
                $parsedArgs = $subCommand->parseArgs($args, $currentParameter);
            }
        } catch (FlagParseException|NoArgumentFoundException) {
            $this->sendUsageMessage($sender, $subCommand);
            return;
        } catch (ArgumentParseException) {
            if ($currentParameter?->getCustomErrorMessage() !== null) $sender->warn($currentParameter->getCustomErrorMessage());
            else $this->sendUsageMessage($sender, $subCommand);
            return;
        }

        $finalFlags = array_merge($commandGlobalFlags, $subCommand === null ? $commandRegularFlags : $subCommandFlags);

        if ($subCommand === null) {
            if (!$this->run($sender, $label, $parsedArgs ?? $args, null, $finalFlags)) {
                $this->sendUsageMessage($sender);
            }
        } else {
            if (!($this->useRegularHandlerForSubCommands ? $this->run($sender, $label, $parsedArgs ?? $args, $subCommand, $finalFlags) : $subCommand->run($sender, $label, $parsedArgs ?? $args, $finalFlags))) {
                $this->sendUsageMessage($sender, $subCommand);
            }
        }
    }

    abstract public function run(ICommandSender $sender, string $label, array $args, ?SubCommand $subCommand, array $flags): bool;

    private function buildUsageMessage(?SubCommand $subCommand = null): string {
        if ($subCommand === null) {
            $usage = "";
            $lastSubCommand = array_key_last($this->subCommands);
            foreach ($this->subCommands as $name => $subCommand) {
                $usage .= $subCommand->getUsage();
                if ($lastSubCommand !== $name) $usage .= "\n";
            }

            if ($lastSubCommand !== null && count($this->parameters) > 0) $usage .= "\n";
            if ($usage == "") $usage .= $this->getName();
            foreach ($this->parameters as $parameter) {
                $usage .= $parameter->isOptional() ?
                    " [" . $parameter->getName() . ": " . $parameter->getType() . "]" :
                    " <" . $parameter->getName() . ": " . $parameter->getType() . ">";
            }

            foreach ($this->flags as $flag) {
                $usage .= " " . $flag->buildUsage();
            }
        } else {
            $usage = $subCommand->getUsage();
        }

        return $usage;
    }

    public function sendUsageMessage(ICommandSender $sender, ?SubCommand $subCommand = null, ?CloudLogLevel $logLevel = null): void {
        foreach (explode("\n", $this->getUsage($subCommand)) as $line) {
            $sender->log($logLevel ?? CloudLogLevel::WARN(), trim($line));
        }
    }

    public function registerSubCommand(SubCommand $subCommand): self {
        $this->subCommands[strtolower($subCommand->getName())] = $subCommand;
        return $this;
    }

    public function getSubCommand(string $name): ?SubCommand {
        $name = strtolower($name);
        if (isset($this->subCommands[$name])) return $this->subCommands[$name];
        return array_find($this->subCommands, fn(SubCommand $subCommand) => in_array($name, $subCommand->getStandaloneAliases()));
    }

    public function mustUseSubCommands(): int {
        return count($this->subCommands) > 0 && count($this->parameters) == 0;
    }

    public function getSubCommands(): array {
        return $this->subCommands;
    }

    public function isUseRegularHandlerForSubCommands(): bool {
        return $this->useRegularHandlerForSubCommands;
    }

    public function getName(): string {
        return $this->name;
    }

    public function getDescription(): string {
        return $this->description;
    }

    public function getUsage(?SubCommand $subCommand = null): string {
        if ($subCommand !== null) return $this->buildUsageMessage($subCommand);
        return $this->usage ?? $this->buildUsageMessage();
    }

    public function getAliases(): array {
        return $this->aliases;
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/console/command/Command.java
package de.pocketcloud.cloud.console.command;

import de.pocketcloud.api.logging.CloudLogLevel;
import de.pocketcloud.cloud.PocketCloud;
import de.pocketcloud.cloud.console.command.ctx.CommandContext;
import de.pocketcloud.cloud.console.command.desc.CommandDescription;
import de.pocketcloud.cloud.console.command.exception.ArgumentParseException;
import de.pocketcloud.cloud.console.command.exception.FlagParseException;
import de.pocketcloud.cloud.console.command.exception.NoArgumentFoundException;
import de.pocketcloud.cloud.console.command.flag.CommandFlag;
import de.pocketcloud.cloud.console.command.holder.CommandUtilityHolder;
import de.pocketcloud.cloud.console.command.parameter.BaseCommandParameter;
import de.pocketcloud.cloud.console.command.sender.CommandSender;
import de.pocketcloud.cloud.console.command.sub.SubCommand;
import de.pocketcloud.common.concurrent.Promise;
import lombok.Getter;
import org.jetbrains.annotations.Nullable;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

@Getter
public abstract class Command extends CommandUtilityHolder {

    private final String name;
    private final String description;
    private final String usage;
    private final String[] aliases;

    private final Map<String, SubCommand> subCommands = new HashMap<>();
    
    public Command() {
        if (!this.getClass().isAnnotationPresent(CommandDescription.class)) {
            throw new IllegalStateException(this.getClass().getSimpleName() + " requires @CommandDescription");
        }

        CommandDescription annotation = getClass().getAnnotation(CommandDescription.class);
        name = annotation.name();
        usage = annotation.usage().isBlank() ? null : annotation.usage();
        description = annotation.description();
        aliases = annotation.aliases();

        prepare();
    }

    public Promise<Boolean> awaitConfirmation(CommandSender sender, String prompt) {
        return awaitConfirmation(sender, prompt, 10);
    }

    public Promise<Boolean> awaitConfirmation(CommandSender sender, String prompt, int timeoutInSeconds) {
        return PocketCloud.instance().commands().awaitConfirmation(this, sender, prompt, timeoutInSeconds);
    }

    abstract public void prepare();

    public void handle(CommandSender sender, String label, List<String> args) {
        ScanResult commandFlags;
        try {
            commandFlags = scanAndCleanFlags(args);
        } catch (FlagParseException e) {
            sendUsageMessage(sender);
            return;
        }

        SubCommand subCommand = null;
        if (!subCommands.isEmpty()) {
            if (mustUseSubCommands() && args.isEmpty()) {
                sendUsageMessage(sender);
                return;
            }

            if (!args.isEmpty()) subCommand = getSubCommand(args.getFirst());

            if (subCommand == null && mustUseSubCommands()) {
                sendUsageMessage(sender);
                return;
            } else if (subCommand != null) {
                args.removeFirst();
            }
        }

        Map<String, Object> subCommandFlags = new LinkedHashMap<>();
        LastParsed currentParameter = new LastParsed();
        Map<String, Object> parsedArgs;
        try {
            if (subCommand == null) {
                parsedArgs = parseArgs(args, currentParameter);
            } else {
                ScanResult subFlags = subCommand.scanAndCleanFlags(args, true);
                subCommandFlags = subFlags.regularFlags();
                parsedArgs = subCommand.parseArgs(args, currentParameter);
            }

        } catch (FlagParseException | NoArgumentFoundException e) {
            sendUsageMessage(sender, subCommand);
            return;
        } catch (ArgumentParseException e) {
            if (currentParameter.value != null && currentParameter.value.getCustomErrorMessage() != null) {
                sender.warn(currentParameter.value.getCustomErrorMessage());
            } else {
                sendUsageMessage(sender, subCommand);
            }

            return;
        }

        Map<String, Object> finalFlags = new LinkedHashMap<>(commandFlags.globalFlags());
        if (subCommand == null) {
            finalFlags.putAll(commandFlags.regularFlags());
        } else {
            finalFlags.putAll(subCommandFlags);
        }

        boolean success;
        if (subCommand == null) {
            success = run(sender, new CommandContext(label, parsedArgs, finalFlags));
        } else {
            success = subCommand.run(sender, new CommandContext(label, parsedArgs, finalFlags));
        }

        if (!success) {
            sendUsageMessage(sender, subCommand);
        }
    }

    public abstract boolean run(CommandSender sender, CommandContext ctx);

    private String buildUsageMessage(SubCommand subCommand) {
        if (subCommand != null) return subCommand.getUsage();
        StringBuilder usage = new StringBuilder();
        int index = 0;
        int size = subCommands.size();
        for (SubCommand cmd : subCommands.values()) {
            usage.append(cmd.getUsage());
            if (++index < size) usage.append("\n");
        }

        if (!subCommands.isEmpty() && !parameters.isEmpty()) {
            usage.append("\n");
        }

        if (!mustUseSubCommands()) {
            usage.append(getName());

            for (BaseCommandParameter parameter : parameters) {
                usage.append(parameter.isOptional()
                        ? " [" + parameter.getName() + ": " + parameter.getType() + "]"
                        : " <" + parameter.getName() + ": " + parameter.getType() + ">"
                );
            }

            for (CommandFlag flag : flags.values()) {
                usage.append(" ").append(flag.buildUsage());
            }
        }

        return usage.toString();
    }

    public void sendUsageMessage(CommandSender sender, SubCommand subCommand, CloudLogLevel logLevel) {
        for (String line : getUsage(subCommand).split("\n")) {
            sender.log(logLevel != null ? logLevel : CloudLogLevel.WARN, line.trim());
        }
    }

    public void sendUsageMessage(CommandSender sender) {
        sendUsageMessage(sender, null, null);
    }

    public void sendUsageMessage(CommandSender sender, SubCommand subCommand) {
        sendUsageMessage(sender, subCommand, null);
    }

    public Command registerSubCommand(SubCommand subCommand) {
        subCommands.put(subCommand.getName().toLowerCase(), subCommand);
        return this;
    }

    @Nullable
    public SubCommand getSubCommand(String name) {
        return subCommands.get(name.toLowerCase());
    }

    public boolean mustUseSubCommands() {
        return !subCommands.isEmpty() && parameters.isEmpty();
    }

    public String getUsage(SubCommand subCommand) {
        return subCommand != null ? buildUsageMessage(subCommand) : (usage != null ? usage : buildUsageMessage(null));
    }

    public String getUsage() {
        return getUsage(null);
    }
}